### PR TITLE
Add canary_assert to metrics.

### PR DESCRIPTION
**Pull request by willhickey**
#### Summary of Changes
Add a canary_assert macro. 

It checks if the environment variable SOLANA_CANARY_VALIDATOR. If it is set the expression arg is evaluated and causes a panic if false.

### CODE REVIEW

File: metrics/src/metrics.rs [hunk 1]
```diff
@@ -460,6 +460,18 @@ pub fn set_panic_hook(program: &'static str) {
     });
 }
 
+#[macro_export]
+macro_rules! canary_assert {
+    ($assertion:expr) => {
+        let is_canary_validator: bool = env::var("CANARY_VALIDATOR").is_ok();
```
Comment (t-nelson) [inline metrics/src/metrics.rs:466]:
```suggestion
        let is_canary_validator: bool = env::var("SOLANA_CANARY_VALIDATOR").is_ok();
```

**Review by t-nelson — COMMENTED** [review]
I'm somewhat uneasy with adding something like this to the code base.  Can we back stop it with a CLI arg and/or cargo feature to prevent it from being accidentally enabled?

**Comment by jstarry** [comment]
> I'm somewhat uneasy with adding something like this to the code base. Can we back stop it with a CLI arg and/or cargo feature to prevent it from being accidentally enabled?

What's the risk if it's an env var? This is only easy to use for core devs if we can quickly add it to any crate

**Comment by willhickey** [comment]
@jstarry I think this meets the requirements, but I'm interested in any feedback.

I switched from a macro to a function because I was having trouble getting the macro to interact with IS_CANARY_VALIDATOR. This is likely just my inexperience with Rust... if you've got suggestions for a better approach I'll redo it.

@t-nelson Are you envisioning a CLI arg in addition to the env var as a double check? It seems unlikely that the current implementation will lead to problems since 1) it has to be enabled via env var and 2) the asserts shouldn't be failing anyway and if they do we should catch those problems in the canaries.

**Comment by mvines** [comment]
> I think this meets the requirements, but I'm interested in any feedback.

I'm generally confused about what this would be used for.  At minimum there should be a write-up somewhere that talks about what the heck we'd use this for, ideally in the code for easy future reference.  It's not immediately obvious to me from just looking at the patch

File: metrics/src/metrics.rs [hunk 2]
```diff
@@ -460,6 +460,31 @@ pub fn set_panic_hook(program: &'static str) {
...
+    static ref IS_CANARY_VALIDATOR: Arc<RwLock<bool>> = Arc::new(RwLock::new(false));
+}
+
+pub fn set_is_canary_validator(is_canary_validator: bool) {
+    info!("Setting IS_CANARY_VALIDATOR: {}", is_canary_validator);
+    *IS_CANARY_VALIDATOR.write().unwrap() = is_canary_validator;
+}
+
+pub fn canary_assert(assertion: bool, failure_message: &str) {
+    let is_canary_validator = *IS_CANARY_VALIDATOR.read().unwrap();
```
Comment (mvines) [inline metrics/src/metrics.rs:473]: Looks like `set_is_canary_validator()` can disappear if a `std::sync::Once` is employed in here to read the env var

**Comment by willhickey** [comment]
canary validator - a validator that runs a not-yet-released version of the validator software with the intent of catching bugs that have been merged but not released. Solana Labs currently operates several canary validators (bvb, bve, c1, c2, c3, c4, c5, c6, c7) and I'm working on formalizing how we use them.

The intent of canary_assert is to provide a wrapper around assert! that logs all failures but only panics if the env var SOLANA_CANARY_VALIDATOR is set. This would allow more aggressive use of assert to confirm assumptions in code without fear of panicking the majority of validators.

The value depends on whether core contributors find themselves not using assert because they're not sure that an assumption will be true 100% off the time, or simply hesitating to use assert because the failure behavior is too drastic. I'll defer to devs with more Solana experience whether this provides enough value. 

If there's interested in this functionality I will clean up the code and add a doc comment.

@jstarry Feel free to expand/correct this if I haven't correctly described what you envisioned.

**Comment by jstarry** [comment]
> 🤔 this feels a little off to me. An assert should be an assert. If there's code we aren't confident about, it has no place landing on master to begin with. Goal: master is always in a shippable state (which I admit we could do better about)

I think in the past we've had the mentality that there are certain invariants that should always hold and are worth asserting. Capitalization checks come to mind there. But there are certain less critical invariants that probably aren't worth stopping the cluster for. The dev team should be notified asap about those issues but there's no good way for these to be surfaced right now. Maybe a canary assert isn't the right approach, but it'd be nice to have some alerting infrastructure setup on trusted nodes so that we know immediately when something unexpected is happening. Adding an assertion seemed like a quick and dirty way to do it for now because it's harder to ignore a crashing validator than it is to ignore a random metric datapoint.

**Comment by mvines** [comment]
I totally agree with the direction. Building better alerting around error datapoints feels better to me, and there's no reason why those couldn't be really yelly.  

A crashing node sucks though.  Some cases that come to mind immediately are:
1. A crash on Saturday morning! The person that cares about the particular event may not get to it until Monday or worse, meanwhile we're getting no further test coverage from that build and somebody now needs to recover potentially multiple machines.
2. You want to monitor something but somebody else's canary is currently crashing your node. You need to now manually back out their `canary_assert` or be blocked until they do something about it.

**Comment by t-nelson** [comment]
Can't we just deploy a one-off build like we always have for this kind of stuff?

**Comment by jstarry** [comment]
> Can't we just deploy a one-off build like we always have for this kind of stuff?

If there's something with an immediate concern sure, but there's also value in passing alerting (emphasize the canary analogy here) where something invalidates the invariant in the future or some rarely seen condition occurs on a cluster. Epoch boundaries come to mind here.

**Comment by t-nelson** [comment]
So the plan is to leave these in long term?  I think I'm with mvines, if we aren't confident enough for full assert, it doesn't belong in master.  We're just going to collect these things until one kills MB

**Comment by mvines** [comment]
Moar alert tooling surrounding `datapoint_error!()` is the way IMO

**Pull request closed without merging**